### ISSUE TITLE
C++ APIs - "The size of the buffer provided for a bound variable is invalid"

### ISSUE THREAD

**Issue by tgolsson** [microsoft/Windows-Machine-Learning#16]
Hello,

I'm interested in using the pure C++ APIs, as described [here](https://msdn.microsoft.com/en-us/library/windows/desktop/mt845846(v=vs.85).aspx#). However, these are very poorly documented, which is making it very hard to use them. Aside from my error, it'd be very helpful if you could also provide examples for using this API without having to use C++/CX as provided by `MLGEN.exe`.

As of now, I've got a very simple test which works for creating a context and runtime, loading a model, and binding values. All of these return `S_OK` when called. However, the `HRESULT` returned from `EvaluateModel(context)` has the following error description:

>  The size of the buffer provided for a bound variable is invalid. 

However, given the network architecture I expect to give the same size data in as out, and both are sized accordingly. Both the input and output report being `TENSOR`s, and I've set `pShape` to the corresponding to the input/output shape, except replacing `[-1, ...]` with `[1, ...]` for the output . `pData` is bound to an appropriately sized vector, which has been resized to the exact size of both input and output (4096 floats, 16384 bytes). I set the `DataSize` member of the binding to 16384, as well. 

This is the output from `EnumerateModelInput`/`...Output`:

```text
Input #1
        Name: input_1:0
        Desc:
        FeatureType: 1 // Tensor
        ElemType: 1 // Float
        Dims: 4
        Shape: [ 1, 1, 64, 64, ]
Output #1
        Name: conv2d_8/Relu:0
        Desc:
        FeatureType: 1 // Tensor
        ElemType: 1 // Float
        Dims: 4
        Shape: [ -1, 64, 64, 1, ]
```

I sadly cannot share much more publicly, but I'll happily supply debug info or help out via e-mail.

**Comment by tgolsson** [comment]
Some more digging, and I think it's related to this issue MicrosoftDocs/windows-uwp#463, which has been closed but not solved. 

For what it's worth, I can move the error between the `context->BindValue(...)` and `runtime->EvaluateModel(...)` depending on whether I set batchsize to `1` or `-1`. If I leave it as negative (per the output description), it'll give me the same error, but at bind time. If I "fix" it and make it positive, it'll instead error out during evaluation. 

It seems like there's some interaction with batch shapes and datasizes that doesn't go down well. I noticed that `DataSize` of a tensorbinding is an unsigned integer, but at surface it seems like something somewhere goes `int bytesExpected = product(pShape);`, which returns a negative value and things get wonky.

**Comment by tgolsson** [comment]
To further follow up, I manually edited my ONNX graph file, and forced the `-1` to `1`, and it gives me the same issue.

**Comment by youngkim93** [comment]
Hi Tom, I'm sorry for the delay in response. I'm assuming that you are using Preview API in RS4 and want to see if you are hitting the same issue on the insider preview build. Have you by any chance tried your case on the preview build with newer SDK? If not I can give you more information and would love to get more debug information

**Comment by walrusmcd** [comment]
Closing this out as an issue againts the Preview API .    Please try this on the new insider builds, we just released our new samples and MNIST !

**Comment by rribani** [comment]
This issue is still happening, not with the MNIST sample but when I try any ONNX model. Also, the pre-processing tools in .NET are terrible! Came on guys, even in iOS or Android is trivial to do this!!

**Comment by ryanlai2** [comment]
Hi @rribani can I get more information on your issue? Have you tried running your ONNX model with [WinMLRunner](https://github.com/microsoft/Windows-Machine-Learning/releases/tag/1.2.0.1)?

**Comment by rribani** [comment]
Hi @ryanlai2, sure. I tried the WinMLRunner but it fails since my model input is float and has a fixed size. I'm using a model based on MobileNet (input 224x224) After a while trying, I had to change the model input type to TensorFloat and  I struggled for days to figure out how to resize the content from the stream.
I had to use a BitmapDecoder, a BitmapTransform, a PixelDataProvider, DetachPixelData(), a WriteableBitmap, a new SoftwareBitmap, a buffer and then create a TensorFloat from that. Ufff!

Here is my code btw:
```
                SoftwareBitmap softwareBitmap;
                using (IRandomAccessStream stream = await selectedStorageFile.OpenAsync(FileAccessMode.Read))
                {
                    // Create the decoder from the stream 
                    BitmapDecoder decoder = await BitmapDecoder.CreateAsync(stream);

                    BitmapTransform transform = new BitmapTransform() { ScaledHeight = 224, ScaledWidth = 224 };
                    PixelDataProvider pixelData = await decoder.GetPixelDataAsync(
                        BitmapPixelFormat.Rgba8,
                        BitmapAlphaMode.Premultiplied,
                        transform,
                        ExifOrientationMode.RespectExifOrientation,
                        ColorManagementMode.DoNotColorManage);

                    byte[] pixels = pixelData.DetachPixelData();

                    //Initialize writable bitmap.
                    var wBitmap = new WriteableBitmap((int)decoder.PixelWidth, (int)decoder.PixelHeight);
                    var wb_stream = wBitmap.PixelBuffer.AsStream();
                    await wb_stream.WriteAsync(pixels, 0, pixels.Length);
                    softwareBitmap = SoftwareBitmap.CreateCopyFromBuffer(wBitmap.PixelBuffer, BitmapPixelFormat.Bgra8, 224, 224, BitmapAlphaMode.Premultiplied);

                    IEnumerable<long> shape = new long[] { 1, 224, 224, 3 };
                    var pixelsFloat = new float[pixels.Length / 4];
                    System.Buffer.BlockCopy(pixels, 0, pixelsFloat, 0, pixels.Length);
                    TensorFloat.CreateFromArray(shape, pixelsFloat);
                    modelInput.input_1_01 = TensorFloat.CreateFromArray(shape, pixelsFloat);

                }
```
This is working for inference in the model, but the results are not matching with the outputs from keras/tensorflow. I know there are some pre-processing steps missing, like normalization and ensure channel order. Probably I'll spend some more days to figure out how to do that, since the samples does not help. 
Seriously, if you want people to use WindowsML, you must simplify this! It's not intuitive at all (even with documentation)!
Thanks,
Ribani

**Comment by LPBourret** [comment]
Hi Ribani,
Let me try to help:
as you may know WindowsML makes it pretty easy to automatically tensorize image input in NCHW format by simply [binding an ImageFeatureValue instance](https://github.com/microsoft/Windows-Machine-Learning/blob/b803e660222bca0c0121df3dbd54c66021de31a7/Samples/SqueezeNetObjectDetection/UWP/cs/MainPage.xaml.cs#L194) which internally resizes to smallest axis then crops to accomodate prescribed aspect ratio and tensorizes to proper pixel order. In your case it seems your input shape is instead NHWC and demands a bit more work (i.e. [custom tensorization](https://github.com/Microsoft/Windows-Machine-Learning/tree/master/Samples/CustomTensorization)). 

- First thing to understand is: in your reference framework, how are you pre-processing images in terms of size (crop, rescale). 
  - In your example code above, you could use instead the VideoFrame API to [encapsulate the retrieved SoftwareBitmap](https://docs.microsoft.com/en-us/uwp/api/windows.media.videoframe.createwithsoftwarebitmap#Windows_Media_VideoFrame_CreateWithSoftwareBitmap_Windows_Graphics_Imaging_SoftwareBitmap_) then use the [VideoFrame.CopyToAsync()](https://docs.microsoft.com/en-us/uwp/api/windows.media.videoframe.copytoasync#Windows_Media_VideoFrame_CopyToAsync_Windows_Media_VideoFrame_) to copy its content to a 224x224 staging VideoFrame instance. Note that you [can specify the BitmapBounds](https://docs.microsoft.com/en-us/uwp/api/windows.media.videoframe.copytoasync#Windows_Media_VideoFrame_CopyToAsync_Windows_Media_VideoFrame_Windows_Foundation_IReference_Windows_Graphics_Imaging_BitmapBounds__Windows_Foundation_IReference_Windows_Graphics_Imaging_BitmapBounds__) to use in that last method if you prefer to copy only portion of the original image (crop).
```csharp
SoftwareBitmap softwareBitmap;
using (IRandomAccessStream stream = await selectedStorageFile.OpenAsync(FileAccessMode.Read))
{
  // Create the decoder from the stream 
  BitmapDecoder decoder = await BitmapDecoder.CreateAsync(stream);

  // Get the SoftwareBitmap from the image file
  softwareBitmap = await decoder.GetSoftwareBitmapAsync();
}
// Encapsulate the image within a VideoFrame
VideoFrame inputImage = VideoFrame.CreateWithSoftwareBitmap(softwareBitmap);

// Create a staging frame of the desired size
VideoFrame stagingImage = new VideoFrame(BitmapPixelFormat.Bgra8, 224, 224, BitmapAlphaMode.Premultiplied);

// Copy input image into staging frame. If aspect ratio does not match, this method will pad and add black bar on both side
await inputImage.CopyToAsync(stagingImage);
```
- With your SoftwareBitmap encapsulated in a VideoFrame in hand, second thing has to do with expected pixel values, pixel order and tensorization. 
  - Currently you have an image in BGRA [0,255]. If you wish to normalize it for example to [0,1] float, you can do so at the same time you will create your tensor. 
  - Given your shape, you may only need to drop the A pixel component (i.e. you want a tensor in BGR-BGR-BGR...) or additionaly reorder to RGB (make sure this matches your reference frmaework). Your example code does a block copy that includes the A component. 
  - For creating an array from your image, you can refer to the logic in the [TensorizationHelper sample](https://github.com/microsoft/Windows-Machine-Learning/blob/b803e660222bca0c0121df3dbd54c66021de31a7/Samples/CustomTensorization/CustomTensorization/TensorConvertor.cpp#L77) and adjust it to your need.

Let me know if that helps :)

**Comment by rribani** [comment]
Hi @LPBourret,

The first thing I did was to try the SqueezeNet sample, that uses the VideoFrame. But I need to resize the image to 224x224 (not crop), so here my challenge begins. The only way that I found was using the BitmapTransform (and some more objects).

The TensorFloat class implements the ITensor and I can't do much with it. Any pre-processing or post-processing requires me to move the data to a byte array and work manually with it. To convert from NCHW to NHWC, from BRG to RGB, etc..

I'm trying with this PixelData.. Probably the CustomTensorization sample will help me some way.. I'll post my results here.

Thanks,
Ribani

**Comment by LPBourret** [comment]
Hi @rribani,
as a side note related to your pipeline, you can access directly the pixel buffer underneath a *TensorFloat* via the [CreateReference() method](https://docs.microsoft.com/en-us/uwp/api/windows.ai.machinelearning.tensorfloat.createreference#Windows_AI_MachineLearning_TensorFloat_CreateReference). Just like you [would do with a *SoftwareBitmap*](https://docs.microsoft.com/en-us/windows/uwp/audio-video-camera/imaging#create-or-edit-a-softwarebitmap-programmatically), if you have already bound that *TensorFloat* instance, you can alter it's pixel buffer in-between inference and not rebind it (**as long as the inference device is CPU**, otherwise you need to rebind to let WinML appropriately transfer your data to a DirectX type such as a texture). This allow you to avoid a staging copy or "move" to a staging pixel buffer every time.

In terms of resizing your input image, you probably have good reason for doing that but I would advise others to avoid resizing if possible and instead train or use cropped images to accommodate the wide range of aspect ratio in image capture devices. In my experience, resizing arbitrary aspect ratio to 1:1 may disrupt the details of the original image to a point that affects accuracy on inference.

**Comment by LPBourret** [comment]
closing, feel free to reopen
